### PR TITLE
Rectify: Renderer Extraction and Figure-Spec Schema Contracts

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

Three bugs in the bundle-local-report renderer (BUG-R1: mermaid asset path resolution failure, BUG-R2: figure-spec fence syntax ambiguity, BUG-R3: `image_path` never populated by upstream skills) share a common architectural root: the renderer is embedded as a Python code block inside `SKILL.md` markdown, which means it escapes all static analysis (mypy, ruff, arch tests, the `Path(__file__).parent` prohibition) and the `yaml:figure-spec` schema it consumes has no machine-readable contract shared with producer skills. Additionally, `generate-report/SKILL.md` contains unresolved git merge conflict markers (lines 230-239) that no test or hook detects.

This plan extracts the renderer to a proper Python module, replaces `Path(__file__)` with `pkg_root()` (from `core.paths`) for asset resolution, introduces a shared `FigureSpec` schema, and wires `generate-report` to populate `image_path` after rendering figures. It also resolves the merge conflict in `generate-report/SKILL.md`.

## Conflict Resolution Decisions

The following files had merge conflicts that were automatically resolved.

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/remediation-20260509-200800-376399/.autoskillit/temp/rectify/rectify_renderer_extraction_and_schema_contracts_2026-05-09_210000_part_a.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | Model | count | uncached | output | cache_read | peak_ctx | turns | cache_write | time |
|------|-------|-------|----------|--------|------------|----------|-------|-------------|------|
| rectify | claude-sonnet-4-6 | 1 | 1.2k | 17.1k | 846.9k | 97.9k | 195 | 69.4k | 13m 2s |
| dry_walkthrough | claude-opus-4-6 | 2 | 4.9k | 30.0k | 3.8M | 82.8k | 244 | 135.0k | 14m 51s |
| implement* | MiniMax-M2.7-highspeed | 2 | 11.8M | 68.8k | 6.5M | 73.8k | 517 | 233.1k | 33m 27s |
| prepare_pr* | MiniMax-M2.7-highspeed | 1 | 187.5k | 5.0k | 378.5k | 29.1k | 28 | 41.7k | 1m 44s |
| compose_pr* | MiniMax-M2.7-highspeed | 1 | 52.5k | 1.9k | 198.4k | 28.8k | 19 | 15.2k | 48s |
| review_pr | claude-sonnet-4-6 | 3 | 328 | 102.6k | 1.9M | 101.7k | 124 | 265.8k | 25m 36s |
| resolve_review | claude-sonnet-4-6 | 3 | 1.3k | 106.0k | 12.4M | 128.2k | 362 | 298.6k | 45m 52s |
| ci_conflict_fix | claude-sonnet-4-6 | 3 | 377 | 13.1k | 1.8M | 55.0k | 110 | 108.6k | 4m 15s |
| diagnose_ci* | MiniMax-M2.7-highspeed | 1 | 51.9k | 1.3k | 144.1k | 28.8k | 12 | 41.1k | 41s |
| resolve_ci | claude-sonnet-4-6 | 1 | 240 | 12.6k | 1.4M | 64.2k | 71 | 51.2k | 10m 39s |
| **Total** | | | 12.1M | 358.3k | 29.4M | 128.2k | | 1.3M | 2h 30m |

\* *Step used a non-Anthropic provider; caching behavior may differ.*

## Token Efficiency

| Step | LoC Changed | cache_read/LoC | cache_write/LoC | output/LoC |
|------|-------------|----------------|-----------------|------------|
| rectify | 0 | — | — | — |
| dry_walkthrough | 0 | — | — | — |
| implement | 992 | 6586.1 | 235.0 | 69.3 |
| prepare_pr | 0 | — | — | — |
| compose_pr | 0 | — | — | — |
| review_pr | 0 | — | — | — |
| resolve_review | 219 | 56567.6 | 1363.6 | 484.2 |
| ci_conflict_fix | 3142 | 558.9 | 34.6 | 4.2 |
| diagnose_ci | 0 | — | — | — |
| resolve_ci | 5 | 285602.8 | 10231.2 | 2528.6 |
| **Total** | **4358** | 6737.5 | 289.0 | 82.2 |
## Model Usage Breakdown

| Model | steps | uncached | output | cache_read | cache_write | time |
|-------|-------|----------|--------|------------|-------------|------|
| claude-sonnet-4-6 | 6 | 3.9k | 178.6k | 8.0M | 516.8k | 1h 4m |
| claude-opus-4-6 | 1 | 80 | 16.7k | 1.0M | 80.6k | 13m 20s |
| MiniMax-M2.7-highspeed | 3 | 1.8M | 17.8k | 1.6M | 119.7k | 6m 23s |

### CODE REVIEW

File: tests/skills_extended/test_bundle_local_report.py [hunk 1]
```diff
@@ -59,18 +29,23 @@ def test_renders_minimal_report(tmp_path: Path) -> None:
...
+    )
+    out_html = research_dir / "report.html"
+    out_html.write_text(html, encoding="utf-8")
 
-    html_path = research_dir / "report.html"
-    assert html_path.exists()
-    html = html_path.read_text()
-    assert "mermaid.initialize" in html
-    assert "Hello world paragraph" in html
+    assert out_html.exists()
```
Comment (Trecek) [inline tests/skills_extended/test_bundle_local_report.py:45]: [warning] tests: test_renders_minimal_report assertions are thin — only checks file existence and two string patterns. Does not assert structural correctness (valid HTML skeleton, no template placeholders left unfilled). Consider adding assertions on the HTML structure.

File: tests/skills_extended/test_bundle_local_report.py [hunk 2]
```diff
@@ -138,30 +94,74 @@ def test_images_inserted_from_figure_spec(tmp_path: Path) -> None:
...
-    html = (research_dir / "report.html").read_text()
-    assert 'src="images/fig-1.png"' in html
-    assert 'alt="Main Results"' in html
+    assert 'src="images/fig-1.png"' in body_html
+    assert 'alt="Main Results"' in body_html
 
 
 def test_html_includes_mermaid_version_comment(tmp_path: Path) -> None:
-    """Rendered HTML contains a <!-- mermaid ... --> version comment."""
+    """Rendered HTML contains a <!-- mermaid {version} --> version comment (not 'unknown')."""
```
Comment (Trecek) [inline tests/skills_extended/test_bundle_local_report.py:113]: [warning] tests: test_html_includes_mermaid_version_comment duplicates the render scaffolding already in test_renders_minimal_report (same _find_mermaid_assets / HTML_TEMPLATE.format / _markdown_to_html sequence). Consider folding the version-comment assertion into test_renders_minimal_report to eliminate redundant boilerplate.

File: tests/skills_extended/test_bundle_local_report.py [hunk 3]
```diff
@@ -138,30 +94,74 @@ def test_images_inserted_from_figure_spec(tmp_path: Path) -> None:
...
 
-    html = (research_dir / "report.html").read_text()
     assert "<!-- mermaid " in html
+    assert mermaid_version != "unknown", (
+        "Mermaid version resolved to 'unknown' — _find_mermaid_assets() failed. "
+        "The renderer must use pkg_root() from core.paths."
+    )
+
+
+def test_renderer_uses_importlib_not_dunder_file() -> None:
```
Comment (Trecek) [inline tests/skills_extended/test_bundle_local_report.py:136]: [warning] tests: test_renderer_uses_importlib_not_dunder_file checks ast.Attribute nodes with .attr == '__file__', but bare __file__ Name nodes (e.g. Path(__file__) where Path wraps a Name, not an Attribute) are not caught. The AST walker must also check ast.Name nodes with id == '__file__' to catch all uses.

File: tests/skills_extended/test_bundle_local_report.py [hunk 4]
```diff
@@ -138,30 +94,74 @@ def test_images_inserted_from_figure_spec(tmp_path: Path) -> None:
...
+    """renderer.py must not use Path(__file__) for asset resolution.
+
+    Path(__file__) resolves to the temp directory when running extracted scripts,
```
Comment (Trecek) [inline tests/skills_extended/test_bundle_local_report.py]: [warning] slop: Docstring says 'when running extracted scripts, causing _find_mermaid_assets() to always return (None, "unknown")' — historical context describing the old subprocess/extraction approach that no longer exists. The renderer is now a proper module. Trim the docstring to describe current behaviour only.

File: tests/arch/test_layer_enforcement.py [hunk 5]
```diff
@@ -827,6 +828,51 @@ def test_no_file_based_path_resolution_outside_paths_module() -> None:
...
+    for skill_dir in skill_dirs:
+        for md_file in skill_dir.rglob("*.md"):
+            text = md_file.read_text()
+            python_blocks = re.findall(r"```python\n(.*?)```", text, re.DOTALL)
+            for block_idx, block in enumerate(python_blocks, 1):
+                try:
+                    tree = ast.parse(block)
+                except SyntaxError:
+                    continue
+                for node in ast.walk(tree):
```
Comment (Trecek) [inline tests/arch/test_layer_enforcement.py:853]: [warning] tests: test_no_dunder_file_in_embedded_python_blocks matches only the specific `Path(__file__).parent` AST pattern. Bare `__file__` references (e.g., `open(__file__)`, plain `Path(__file__)`) are not detected. The test name says 'no dunder file' but the implementation is narrower than the name implies, leaving real violations undetected.

File: tests/contracts/test_generate_report_contracts.py [hunk 6]
```diff
@@ -157,3 +157,27 @@ def test_dockerfile_template_wires_bash_env() -> None:
...
+    """Step 2.5 must instruct writing image_path back to visualization-plan.md.
+
+    bundle-local-report reads image_path from yaml:figure-spec blocks in
+    visualization-plan.md. If generate-report renders figures but never writes
+    image_path back, figure insertion silently fails.
+    """
+    text = SKILL_PATH.read_text()
+
+    # Find Step 2.5 section (between ### Step 2.5 and ### Step 3)
+    assert "### Step 2.5" in text, (
```
Comment (Trecek) [inline tests/contracts/test_generate_report_contracts.py:221]: [warning] tests: test_generate_report_writes_image_path_back_to_viz_plan asserts on static SKILL.md text via string splitting on section headings. Any heading rename or reorder silently breaks the split. Asserting 'image_path' appears in a documentation section is a weak proxy for the actual runtime contract.

File: tests/contracts/test_figure_spec_contracts.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        r"```yaml\n# yaml:figure-spec[^\n]*\n(.*?)```",
+        re.DOTALL,
+    )
+    for m in pattern_b.finditer(text):
+        blocks.append((m.group(1), "yaml (with # yaml:figure-spec comment)"))
+
+    return blocks
+
+
+def _field_names_from_yaml_block(block_text: str) -> frozenset[str]:
```
Comment (Trecek) [inline tests/contracts/test_figure_spec_contracts.py:39]: [warning] cohesion: _field_names_from_yaml_block uses yaml.safe_load() to parse figure-spec YAML blocks in tests, while renderer.py uses manual string splitting for the same logical operation. This parsing asymmetry means the two code paths can silently diverge on edge cases (multi-line values, YAML anchors). The YAML parsing logic should live in a single shared utility.

File: tests/contracts/test_figure_spec_contracts.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    for vis_dir in _VIS_LENS_DIR.glob("vis-lens-*"):
+        if vis_dir.name in composite_skills:
+            continue
+        skill_md = vis_dir / "SKILL.md"
+        if skill_md.exists():
+            skills.append(skill_md)
+    return sorted(skills)
+
+
+_VIS_LENS_SKILLS = _non_composite_vis_lens_skills()
```
Comment (Trecek) [inline tests/contracts/test_figure_spec_contracts.py:67]: [warning] cohesion: _VIS_LENS_SKILLS is computed at module level (module-level I/O via _non_composite_vis_lens_skills()). Under xdist, this runs once per worker during collection. The pattern is asymmetric with other module-level constants — consider making the I/O boundary explicit or using a module-scoped fixture.

File: src/autoskillit/report/renderer.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+_FENCE = "```"
+
+
+def _count_keywords(text: str) -> int:
+    return sum(1 for kw in VALIDATION_KEYWORDS if kw in text)
+
+
+def _extract_mermaid_blocks(text: str) -> list[str]:
+    """Return list of mermaid diagram source strings from a markdown file."""
+    return re.findall(rf"{_FENCE}mermaid\r?\n(.*?)\r?\n?{_FENCE}", text, re.DOTALL)
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:78]: [warning] bugs: Regex `rf"{_FENCE}mermaid\r?\n(.*?)\r?\n?{_FENCE}"` — the trailing `\r?\n?` before the closing fence is optional (matches zero chars), creating a boundary ambiguity: the non-greedy `.*?` with DOTALL can stop at the wrong position in malformed input, potentially over-matching or missing closing fences when content immediately precedes them without a trailing newline.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: changes_requested

File: src/autoskillit/report/renderer.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+
+def _validate_diagram_paths(paths_str: str) -> list[str]:
+    """Return mermaid source strings for diagrams with ≥2 validation keywords."""
+    validated: list[str] = []
+    if not paths_str.strip():
+        return validated
+    for raw in paths_str.split(","):
+        p = Path(raw.strip())
+        if not p.exists():
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:88]: [warning] defense: Non-existent diagram path is silently skipped with no warning. A missing path may indicate a caller bug (wrong CWD, stale manifest). Log the skip or collect skipped paths so the caller can detect data loss.

File: src/autoskillit/report/renderer.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        if _count_keywords(content) >= 2:
+            blocks = _extract_mermaid_blocks(content)
+            validated.extend(blocks)
+    return validated
+
+
+def _parse_figure_specs(viz_plan_path: str) -> list[FigureSpec]:
+    """Parse yaml:figure-spec blocks from visualization-plan.md."""
+    specs: list[FigureSpec] = []
+    if not viz_plan_path or not Path(viz_plan_path).exists():
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py]: [warning] defense: _parse_figure_specs calls Path(viz_plan_path).exists() then Path(viz_plan_path).read_text() as two separate Path constructions. Assign to a variable once and reuse — avoids the TOCTOU window and the redundant construction.

File: src/autoskillit/report/renderer.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        return specs
+    text = Path(viz_plan_path).read_text(encoding="utf-8")
+    raw_blocks = re.findall(rf"{_FENCE}yaml:figure-spec\r?\n(.*?)\r?\n?{_FENCE}", text, re.DOTALL)
+    for block in raw_blocks:
+        spec: dict[str, Any] = {}
+        for line in block.splitlines():
+            if ":" in line:
+                k, _, v = line.partition(":")
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:112]: [warning] bugs: Manual YAML parsing via `line.partition(':')` silently truncates values containing colons (e.g., URLs like 'https://example.com' become 'https'). The image_path field may contain colon-bearing paths, causing _insert_images to receive a truncated path and silently skip or misrender images. Use yaml.safe_load() for parsing.

File: src/autoskillit/report/renderer.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                spec[k.strip()] = v.strip()
+        if spec:
+            specs.append(cast(FigureSpec, spec))
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:115]: [critical] defense: cast(FigureSpec, spec) performs no runtime type checking. The dict produced by manual YAML parsing may contain incorrect types (non-string values from mis-parsed lines). A malformed yaml:figure-spec block silently produces an invalid FigureSpec that causes subtle failures in _insert_images. Validate at minimum that expected fields are strings before casting.

File: src/autoskillit/report/renderer.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+def _insert_images(html_body: str, specs: list[FigureSpec]) -> str:
+    """Insert <img> tags after heading matches for each figure-spec."""
+    for spec in specs:
+        section = spec.get("report_section", "")
+        title = spec.get("figure_title", "")
+        img_path = spec.get("image_path", "")
+        if not section or not img_path:
+            continue
+        img_tag = f'<img src="{_html.escape(img_path)}" alt="{_html.escape(title)}">'
+        pattern = rf"(<h[1-6][^>]*>[^<]*{re.escape(section)}[^<]*</h[1-6]>)"
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py]: [warning] bugs: Heading regex `(<h[1-6][^>]*>[^<]*{re.escape(section)}[^<]*</h[1-6]>)` uses [^<]* before/after the section text, which prevents matching headings with nested HTML tags (e.g., <h2><strong>Section</strong></h2>). markdown-it-py produces such nested tags for bold/italic headings. _insert_images silently skips figures for any heading containing inline markup.

File: src/autoskillit/report/renderer.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+def _markdown_to_html(md_text: str) -> str:
+    """Convert markdown to HTML using markdown-it-py."""
+    try:
+        from markdown_it import MarkdownIt  # type: ignore[import]
+
+        md = MarkdownIt()
+        rendered = md.render(md_text)
+        rendered = re.sub(
+            r'<pre><code class="language-mermaid">(.*?)</code></pre>',
+            lambda m: f'<pre class="mermaid">{_html.unescape(m.group(1))}</pre>',
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:148]: [warning] bugs: _html.unescape(m.group(1)) correctly restores raw mermaid source, but if the original markdown contained literal HTML entities (not introduced by escaping), they are unescaped and injected into <pre class="mermaid">, potentially as raw HTML in user-supplied content.

File: src/autoskillit/report/renderer.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    mermaid_dir = pkg_root() / "assets" / "mermaid"
+    js_path = mermaid_dir / "mermaid.min.js"
+    if js_path.exists():
+        ver_file = mermaid_dir / "VERSION"
+        version = ver_file.read_text().strip() if ver_file.exists() else "unknown"
+        return js_path, version
+    return None, "unknown"
+
+
+def main() -> None:
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:169]: [warning] defense: main() is missing a return type annotation (-> None). As a public entry point it should be annotated to make the boundary explicit.

File: src/autoskillit/report/renderer.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+    if len(sys.argv) < 3:
+        sys.stdout.write("html_path = \n")
+        sys.stdout.flush()
+        sys.exit(0)
+
+    research_dir = Path(sys.argv[1])
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:175]: [warning] defense: research_dir is never validated to exist before use as the output directory (lines 196-200). atomic_write or shutil.copy2 will raise FileNotFoundError with no contextual message. Add an early existence check or automatic mkdir.

File: src/autoskillit/report/renderer.py [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+        sys.stdout.write("html_path = \n")
+        sys.stdout.flush()
+        sys.exit(0)
+
+    validated_diagrams = _validate_diagram_paths(all_diagram_paths)
+    mermaid_section = "\n".join(
+        f'<pre class="mermaid">{_html.escape(src)}</pre>' for src in validated_diagrams
+    )
+    specs = _parse_figure_specs(viz_plan_path)
+    md_text = report_path.read_text(encoding="utf-8")
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:191]: [warning] defense: report_path.read_text(encoding='utf-8') is called without error handling. Unreadable files (permission errors, non-UTF-8 content) propagate with no context about which file failed.

File: src/autoskillit/report/renderer.py [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+    body_html = _markdown_to_html(md_text)
+    if specs:
+        body_html = _insert_images(body_html, specs)
+    mermaid_js_src, mermaid_version = _find_mermaid_assets()
+    rendered_html = HTML_TEMPLATE.format(
+        mermaid_version=mermaid_version,
+        mermaid_section=mermaid_section.replace("{", "{{").replace("}", "}}"),
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:198]: [critical] bugs: Double-brace escaping mermaid_section.replace('{','{{').replace('}','}}') corrupts mermaid diagram source: mermaid syntax legitimately uses { } in flowchart node labels, classDiagram definitions, etc. These braces are doubled to {{ }}, breaking mermaid rendering in the browser. Use a non-format-string approach (e.g., string concatenation or Template) to avoid escaping user/diagram content.

File: pyproject.toml [hunk 20]
```diff
@@ -312,6 +320,29 @@ forbidden_modules = [
...
     "autoskillit.server",
     "autoskillit.cli",
+    "autoskillit.report",
+]
+
+# IL-010 | Lens: module-dependency | DS-001
+# Rationale: report/ is an IL-1 package that renders research bundles. It must only
+# import from IL-0 (core) to remain a leaf service that can be tested without
+# spinning up the full autoskillit stack. The renderer is extracted from an
+# embedded SKILL.md code block so it is subject to all static analysis (mypy, ruff,
```
Comment (Trecek) [inline pyproject.toml]: [warning] slop: Comment says 'The renderer is extracted from an embedded SKILL.md code block' — stale after this PR. The renderer is now a proper module in autoskillit.report.renderer. Remove or update this sentence to avoid misleading future readers.

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: 2 critical blocking issues and 18 warnings detected. See inline comments — blocking changes required before merge.

File: src/autoskillit/report/renderer.py [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+    raw_blocks = re.findall(rf"{_FENCE}yaml:figure-spec\r?\n(.*?)\r?\n?{_FENCE}", text, re.DOTALL)
+    for block in raw_blocks:
+        spec: dict[str, Any] = {}
+        for line in block.splitlines():
+            if ":" in line:
+                k, _, v = line.partition(":")
+                spec[k.strip()] = v.strip()
+        if spec:
+            specs.append(cast(FigureSpec, spec))
+    return specs
```
Comment (Trecek) [inline src/autoskillit/report/renderer.py:116]: **Observation from local review round 1:**

[warning] defense: cast(FigureSpec, spec) performs no runtime validation. A malformed or incomplete YAML block silently propagates as a FigureSpec with missing or wrongly-typed fields, deferring failures to call sites instead of raising at construction.

**Evidence:** The renderer already gracefully skips specs missing required fields (if not section or not img_path: continue at line 121-122). Fail-fast at parse time vs. silent skip at use time is a design decision requiring human judgment.

File: src/autoskillit/skills_extended/vis-lens-chart-select/SKILL.md [hunk 22]
```diff
@@ -72,6 +72,7 @@ format: str                  # "svg" | "png" | "pdf"
 target_dpi: int              # 300 for publication, 150 for slides
 library: str                 # "matplotlib" | "seaborn" | "plotly" | "ggplot2" | "vega"
 report_section: str          # section of the paper/report this figure appears in
+image_path: str              # populated by generate-report after rendering; empty until then
```
Comment (Trecek) [inline src/autoskillit/skills_extended/vis-lens-chart-select/SKILL.md:75]: **Observation from local review round 1:**

[warning] cohesion: vis-lens-chart-select/SKILL.md has an inline lifecycle comment on image_path ('populated by generate-report after rendering; empty until then') absent from the other 11 vis-lens SKILL.md files that also added image_path. The field's lifecycle contract documentation is inconsistently localised to one skill.

**Evidence:** All 12 vis-lens SKILL.md files were updated to include image_path in figure-spec blocks; only vis-lens-chart-select has the lifecycle annotation. Decision: propagate to all 12 or remove from chart-select requires human judgment on where lifecycle docs should live.

**Review by Trecek — COMMENTED** [review]
Observations accumulated from 1 local review rounds:

File: src/autoskillit/report/renderer.py (hunk 10 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:88]: Agreed — fixed in 33d4082b. Added sys.stderr.write warning before continue so callers can detect data loss from stale manifests.

File: src/autoskillit/report/renderer.py (hunk 11 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py]: Agreed — fixed in 33d4082b. Assigned viz_plan = Path(viz_plan_path) once; all subsequent reads use this variable.

File: src/autoskillit/report/renderer.py (hunk 14 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py]: Agreed — fixed in 33d4082b. Changed [^<]* to .*? so headings with nested tags like <h2><strong>Section</strong></h2> are matched correctly.

File: src/autoskillit/report/renderer.py (hunk 17 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:175]: Agreed — fixed in 33d4082b. Added research_dir.mkdir(parents=True, exist_ok=True) after path construction.

File: tests/skills_extended/test_bundle_local_report.py (hunk 1 above)
Comment (Trecek) [reply tests/skills_extended/test_bundle_local_report.py:45]: Agreed — fixed in cffa83a0. Added assertions for DOCTYPE, article element, closing html tag, and absence of unfilled template placeholders.

File: tests/skills_extended/test_bundle_local_report.py (hunk 3 above)
Comment (Trecek) [reply tests/skills_extended/test_bundle_local_report.py:136]: Agreed — fixed in cffa83a0. Added ast.Name id=='__file__' check alongside the existing Attribute check.

File: tests/skills_extended/test_bundle_local_report.py (hunk 4 above)
Comment (Trecek) [reply tests/skills_extended/test_bundle_local_report.py]: Agreed — fixed in cffa83a0. Trimmed docstring to describe current behaviour only.

File: tests/arch/test_layer_enforcement.py (hunk 5 above)
Comment (Trecek) [reply tests/arch/test_layer_enforcement.py:853]: Agreed — fixed in 88153d80. Replaced narrow Path(__file__).parent compound check with simple ast.Name id=='__file__' walk.

File: pyproject.toml (hunk 20 above)
Comment (Trecek) [reply pyproject.toml]: Agreed — fixed in 973737a0. Removed the stale sentence; IL-010 rationale now reflects the current module structure.

File: tests/skills_extended/test_bundle_local_report.py (hunk 2 above)
Comment (Trecek) [reply tests/skills_extended/test_bundle_local_report.py:113]: Investigated — this is intentional. test_renders_minimal_report checks render correctness (file exists, body content, HTML structure); test_html_includes_mermaid_version_comment checks the version comment contract. These are distinct invariants — consolidating would make test failures less specific.

File: tests/contracts/test_figure_spec_contracts.py (hunk 8 above)
Comment (Trecek) [reply tests/contracts/test_figure_spec_contracts.py:67]: Investigated — this is intentional. Module-level computation is per-worker under xdist (tests/CLAUDE.md: 'Module-level globals are per-worker (separate processes)'), so no cross-worker state sharing occurs. The pattern is consistent with other module-level parametrize sources in this project.

File: src/autoskillit/report/renderer.py (hunk 9 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:78]: Investigated — this is intentional. The \r?\n? before the closing fence is intentional: it allows matching mermaid blocks whose content ends without a trailing newline (valid CommonMark). The non-greedy .*? prevents over-matching. This was part of the CRLF fix in commit 05fba7e4.

File: src/autoskillit/report/renderer.py (hunk 12 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:112]: Investigated — this is intentional. str.partition(':') splits on the FIRST colon and returns the full tail — 'image_path: https://example.com'.partition(':') gives ('image_path', ':', ' https://example.com'). Values containing subsequent colons are preserved correctly. The reviewer appears to have conflated partition with split.

File: src/autoskillit/report/renderer.py (hunk 13 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:115]: Investigated — this is intentional. All values produced by partition(':') parsing are strings; incorrect-type values cannot arise from this path. The graceful guards at lines 121-122 ('if not section or not img_path: continue') handle missing fields without runtime error.

File: src/autoskillit/report/renderer.py (hunk 15 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:148]: Investigated — this is intentional. _html.unescape is required here: markdown-it-py HTML-escapes code block content (< → &lt;, & → &amp;), so unescape restores the original mermaid source for the browser to render. Without it, mermaid diagrams fail to parse.

File: src/autoskillit/report/renderer.py (hunk 16 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:169]: Investigated — this is intentional. main() already carries -> None annotation (line 165 in the current file). No change needed.

File: src/autoskillit/report/renderer.py (hunk 18 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:191]: Investigated — this is intentional. The existence check at line 176 covers FileNotFoundError. Permission errors and encoding issues propagate with a clear Python traceback naming report_path. Adding a try/except here would obscure the failing path rather than clarify it.

File: src/autoskillit/report/renderer.py (hunk 19 above)
Comment (Trecek) [reply src/autoskillit/report/renderer.py:198]: Investigated — this is intentional. str.format() converts {{ -> { and }} -> } in the rendered output. The replace('{','{{') escaping ensures literal braces in mermaid source and body HTML are not interpreted as format placeholders; the final HTML contains correct single braces. This is standard Python format-string practice.

File: tests/contracts/test_generate_report_contracts.py (hunk 6 above)
Comment (Trecek) [reply tests/contracts/test_generate_report_contracts.py:221]: Valid observation — flagged for design decision. The section-split approach is fragile on heading renames. Flagged for human decision on preferred contract form — options include: regex match, searching only in the section body, or a fixture-based approach.

File: tests/contracts/test_figure_spec_contracts.py (hunk 7 above)
Comment (Trecek) [reply tests/contracts/test_figure_spec_contracts.py:39]: Valid observation — flagged for design decision. The test-side uses yaml.safe_load while renderer.py uses partition. The asymmetry is a cohesion concern. Flagged for human decision on whether renderer should add a yaml dependency for consistency.
